prometheus/histogram.go

```diff
@@ -1835,3 +1835,143 @@ func (n *nativeExemplars) addExemplar(e *dto.Exemplar) {
 		n.exemplars = append(n.exemplars[:nIdx], append([]*dto.Exemplar{e}, append(n.exemplars[nIdx:rIdx], n.exemplars[rIdx+1:]...)...)...)
 	}
 }
+
+type constNativeHistogram struct {
+	desc *Desc
+	dto.Histogram
+	labelPairs []*dto.LabelPair
+}
+
+func validateCount(count uint64, negativeBuckets, positiveBuckets map[int]int64, zeroBucket uint64) error {
+	var bucktCountSum int64
+	for _, v := range positiveBuckets {
+		bucktCountSum += v
+	}
+	for _, v := range negativeBuckets {
+		bucktCountSum += v
+	}
+	bucktCountSum += int64(zeroBucket)
+	if bucktCountSum != int64(count) {
+		return fmt.Errorf("bucket sums do not add up to total count")
```

Comment:
The count of observations can actually be greater than the sum of all bucket counts because NaN observations go into no bucket (but then the sum of observations must be NaN).

So these are the conditions to test:
- If the sum of observations is NaN, the number of observations must be greater or equal to the sum of all bucket counts.
- Otherwise, the number of observations must be equal to the sum of all bucket counts (what you are validating right now).

As a side note, let's be more explicit in the error message. "bucket sums do not add up to total count" could mean many things. How about "the sum of all bucket populations exceeds the count of observations"?